midi_app_controller/gui/midi_status.py

```diff
@@ -125,55 +162,91 @@ def select_controller(name: str) -> None:
         layout.addLayout(self._horizontal_layout("Binds:", self.current_binds))
+        layout.addWidget(self.show_binds_file_button)
+        layout.addWidget(self.copy_binds_button)
+        layout.addWidget(self.delete_binds_button)
+        layout.addWidget(self.edit_binds_button)
         layout.addLayout(self._horizontal_layout("MIDI input:", self.current_midi_in))
         layout.addLayout(self._horizontal_layout("MIDI output:", self.current_midi_out))
         layout.addLayout(status_layout)
-        layout.addWidget(self.edit_binds_button)
         layout.addWidget(self.start_handling_button)
         layout.addWidget(self.stop_handling_button)
         layout.addStretch()
 
         self.setLayout(layout)
 
+        self.refresh()
+
+    def refresh(self):
+        """Updates all widgets to ensure they match the data stored inside the StateManager."""
+
+        self.current_controller.refresh_items()
+        self.current_controller.set_current(state.selected_controller)
+        self.show_controllers_file_button.setEnabled(
+            state.selected_controller is not None
+        )
+
+        self.current_binds.refresh_items()
+        self.current_binds.set_current(state.selected_binds)
+        self.show_binds_file_button.setEnabled(state.selected_binds is not None)
+        self.edit_binds_button.setEnabled(state.selected_binds is not None)
+        self.delete_binds_button.setEnabled(state.selected_binds is not None)
+        self.copy_binds_button.setEnabled(state.selected_binds is not None)
+
+        self.status.setText("Running" if state.is_running() else "Not running")
+        self.start_handling_button.setEnabled(not state.is_running())
+        self.stop_handling_button.setEnabled(state.is_running())
+
     def _horizontal_layout(self, label: str, widget: QWidget) -> QHBoxLayout:
         """Creates horizontal layout consisting of the `label` on the left half\
         and the `widget` on the right half."""
-        layout = QHBoxLayout()
+        layout = QVBoxLayout()
         layout.addWidget(QLabel(label))
         layout.addWidget(widget)
         return layout
 
+    def _copy_binds(self):
+        """Copies the currently selected binds to a new file, and selects that file."""
+        assert state.selected_binds is not None, "No binds selected"
+
+        binds = Binds.load_from(state.selected_binds.path)
+        binds.name += f" ({datetime.datetime.now().isoformat().replace(':', '-')} copy)"
+        new_file = binds.save_copy_to(
+            state.selected_binds.path.with_stem(binds.name), Config.BINDS_USER_DIR
+        )
+        state.select_binds(new_file)
+        self.refresh()
+
+    def _delete_binds(self):
+        """Deletes the file with currently selected binds setup."""
+        assert state.selected_binds is not None, "No binds selected"
+        if is_subpath(Config.BINDS_READONLY_DIR, state.selected_binds.path):
+            raise PermissionError("This config file is read-only")
+
```

Comment:
I would also add an `assert` to check if the path is inside `Config.BINDS_USER_DIR`. Just to feel more assured that a bug won't delete something it shouldn't 😄